Frameworks/CoreLocation/CLGeocoder.mm

```diff
@@ -119,138 +127,114 @@ - (void)reverseGeocodeLocation:(CLLocation*)location completionHandler:(CLGeocod
+            }));
         }
     }
 }
 
 /**
  @Status Interoperable
 */
-- (void)geocodeAddressString:(NSString*)addressString completionHandler:(CLGeocodeCompletionHandler)completionHandler {
-    @synchronized(self) {
-        if (self.isGeocoding) {
-            dispatch_async(dispatch_get_main_queue(), ^{
-                completionHandler(nullptr, [NSError errorWithDomain:@"kCLErrorDomain" code:kCLErrorGeocodeCanceled userInfo:nullptr]);
-            });
-        } else {
-            self.geocoding = true;
-            [WSMMapLocationFinder findLocationsAsync:addressString
-                referencePoint:nullptr
-                success:^void(WSMMapLocationFinderResult* results) {
-                    self.geocoding = false;
+- (void)geocodeAddressDictionary:(NSDictionary*)addressDictionary completionHandler:(CLGeocodeCompletionHandler)completionHandler {
+    const NSString* addressStreet = (const NSString*)[addressDictionary objectForKey:@"Street"];
+    const NSString* addressCity = (const NSString*)[addressDictionary objectForKey:@"City"];
+    const NSString* addressState = (const NSString*)[addressDictionary objectForKey:@"State"];
+    const NSString* addressZIP = (const NSString*)[addressDictionary objectForKey:@"ZIP"];
+    const NSString* addressCountry = (const NSString*)[addressDictionary objectForKey:@"Country"];
+
+    NSMutableString* fullAddress = [[NSMutableString alloc] init];
+    if (addressStreet) {
+        [fullAddress appendFormat:@"%@, ", addressStreet];
+    }
+
+    if (addressCity) {
+        [fullAddress appendFormat:@"%@, ", addressCity];
+    }
 
-                    WSMMapLocationFinderStatus status = results.status;
-                    NSError* geocodeStatus = parseError(results);
+    if (addressState && addressZIP) {
+        [fullAddress appendFormat:@"%@ %@, ", addressState, addressZIP];
+    } else if (addressState) {
+        [fullAddress appendFormat:@"%@, ", addressState];
+    } else if (addressZIP) {
+        [fullAddress appendFormat:@"%@, ", addressZIP];
+    }
 
-                    NSMutableArray* geocodeResult = [[NSMutableArray alloc] init];
-                    createResultsArray(results, geocodeResult);
+    if (addressCountry) {
+        [fullAddress appendFormat:@"%@", addressCountry];
+    }
 
-                    dispatch_async(dispatch_get_main_queue(), ^{
-                        completionHandler(geocodeResult, geocodeStatus);
-                    });
-                }
-                failure:^void(NSError* error) {
-                    self.geocoding = false;
-                    dispatch_async(dispatch_get_main_queue(), ^{
-                        completionHandler(nullptr, error);
-                    });
-                }];
-        }
+    if ([fullAddress hasSuffix:@", "]) {
+        [fullAddress deleteCharactersInRange:NSMakeRange([fullAddress length] - 2, 2)];
     }
+
+    [self _geocodeAddress:fullAddress atPoint:nullptr completionHandler:completionHandler];
```

Comment:
ah...makes sense.

---
In reply to: [111603769](https://github.com/Microsoft/WinObjC/pull/2497#discussion_r111603769) [](ancestors = 111603769)